### ISSUE TITLE
Lua API: Add minetest.register_on_item_register()

### ISSUE THREAD

**Issue by Wuzzy2** [minetest/minetest#4754]
This suggestion is fairly straightforward:

Add a function `minetest.register_on_item_register(func)`. `func` is a callback function.
This function is to be called whenever a new node/tool/craftitem or other item is registered.

`func` is a function with parameters `name, definition` and have basically same meaning in `minetest.register_node`, `minetest.register_tool` and `minetest.register_craftitem`.
`definition` should be the value of the item definition as stored in `minetest.registered_items`, so this means it includes the `type` and `origin` (or whatever they're called) fields.

This function will come in handy when you want to create a mod which wants to modify existing items in bulk without knowing them in advance. This is especially important if you need some function which goes through all existing items. The problem is, this is not really possible. You never truly know when the final item has been registered, so a simple loop through `minetest.registered_items` is sub-obtimal. This will also make it possible to use `minetest.override_item` on items without depending on the origin mod or even knowing the item beforehand.

Why this is needed:
Currently, mods which need to somehow handle all items do this ugly workaround:

    minetest.after(0, function()
        for k,v in pairs(minetest.registered_items) do
            -- Do item-specific modifications or reactions here
        end
    end

The assumption is, that this `minetest.after` “ensures” that all items have been registered. I am not sure, this just doesn't seem right. It seems to work, but it is bad, bad style IMO. This `minetest.after(0` idiom seems to be (sadly) quite common, in my own mods collection I just found about 20 mods which (ab-)use it. :-(

Example mods which loop through all items or a subset of all items:
* `doc_items`: Scans “all” item definitions in order to create help pages
* `real_suffocation`: Loops through “all” nodes to add drowning damage for solid nodes (aka “suffocation”)
* Even `creative` from Minetest Game uses the terrible `minetest.after(0` idiom to loop through the items to add to the creative inventory

**Comment by orwell96** [comment]
I want this. Maybe I'll implement it...

**Comment by rubenwardy** [comment]
Please note that it's currently possible to do this by overriding minetest.register_item. Overriding functions is a little hacky though

**Comment by Wuzzy2** [comment]
I agree, I prefer to not overwrite any core Lua API functions, it can get ugly soon.

**Comment by orwell96** [comment]
Problem: What is with items that are registered before the callback gets registered?

**Comment by Wuzzy2** [comment]
@orwell96: Minetest can just quickly plug in all already existing items into the registered callback, right after it has been registered.

**Comment by sfan5** [comment]
@Wuzzy2 what kind of problems arise when you overwrite core Lua API functions?

**Comment by Wuzzy2** [comment]
@sfan5: It is bad practice IMO. What if not only 1 mod does this, but 10 (just for example)? How can you guarantee this will not quickly degrade into a horrible unmanagable mess? I hope you do not encourage such practice.
Also, I think overwriting any core function is non-standard IMO. It is better IMO to have a standard way to do things, instead of hacking things together. I think this is the whole point of having an API in the first place.

**Comment by sfan5** [comment]
> What if not only 1 mod does this, but 10 (just for example)? 

That's no problem because you need to call the "original" item register function (which may already be a wrapper) anyway.

**Comment by Wuzzy2** [comment]
@sfan5: So are you for or against the proposal?

Also, do you consider overwriting core API functions a “good” and “accepted” practice?

**Comment by sfan5** [comment]
@Wuzzy2 it's not a very clean way to do it, but if you do it correctly it satisfies the use case perfectly.

**Comment by raymoo** [comment]
You would have to write the same wrapping code in every mod then, though, unless somebody makes a library that implements register_on_item_register using @sfan5's method.

**Comment by sorcerykid** [comment]
> Also, I think overwriting any core function is non-standard IMO.

Are you sure? That's how minetest.is_protected( ) is and always has been handled. If a mod wants to implement protection, then it has to manually override it. This is accomplished by means of a wrapper, as described in the official Lua API documentation.

**Comment by paramat** [comment]
Luckily we now have something that is better for your needs, called once when all mods have loaded, instead of on every item register (which is probably less efficient).
https://github.com/minetest/minetest/blob/79e393cf4b5027eb9071df23f4834aac3f3a689f/doc/lua_api.txt#L3576-L3578
And i must admit using 'after 0' is not obvious and a little hacky, so it's good we now have this callback.

**Comment by Wuzzy2** [comment]
Ah, this is good news indeed!
